source/SIL.AppBuilder.Portal.Frontend/src/ui/components/project-table/withTableColumns.tsx

```diff
@@ -0,0 +1,111 @@
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { compose } from 'recompose';
+
+import { setColumnSelection as setColumnSelectionData, Column } from '@store/data';
+
+export interface IProvidedProps {
+  columns?: Column[];
+  selectedColumns?: Column[];
+  updateColumnSelection?: (column) => void;
+  isInSelectedColumns?: (columnId) => boolean;
+  columnWidth: (type?: string) => number;
+}
+
+interface IProps {
+  tableName: string;
+  defaultColumns: Column[];
+}
+
+interface IDataProps {
+  setColumnSelection: (column) => void;
+}
+
+export function withTableColumns(props: IProps) {
+
+  function mapStateToProps({ data }) {
+
+    const { tableName, defaultColumns } = props;
+
+    const selectedColumns = data
+      && data.columnSelections
+      && data.columnSelections[tableName] || defaultColumns;
+
+    return {
+      selectedColumns
+    };
+  }
+
+  function mapDispatchToProps(dispatch) {
+
+    const { tableName } = props;
+
+    return {
+      setColumnSelection: (column) => dispatch(setColumnSelectionData(tableName, column))
+    };
+  }
+
+  return InnerComponent => {
+
+    class WrapperComponent extends React.Component<IProvidedProps & IDataProps> {
+
+      updateColumnSelection = (column) => {
```

Comment:
nice! 🎉

Reply:
So we are in the context of a component here. Possible to extract this logic into your helper lib? Same with `isInSelectedColumns`. Probably a few other things, just to try and keep the components simple.

Reply:
Actually, I moved this code from the visual component into this HOC. I could extract it to a helper function but I'm not going to reuse it anywhere else so I think for now I leave it here where it adds context.